File: Perl/ValueOrRef.pl
use Data::Dumper;

# //////////// 数値 ////////////

# 値渡し

my $val1 = 10;
test1($val1);
print $val1."\n"; # 値は変わっていない

sub test1 {
    my $arg1 = $_[0];
    $arg1 = 100;
}

# 参照渡し

my $val11 = 10;
test11(\$val11); # \を付けて参照渡しにする
print $val11."\n"; # 値が変わっている

sub test11 {
    my $arg1 = $_[0];
    $$arg1 = 100;
}

# //////////// 文字列 ////////////

# 値渡し

my $val2 = "ABCD";
test2($val2);
print $val2."\n"; # 値は変わっていない

sub test2 {
    my $arg1 = $_[0];
    $arg1 = "EFG";
}

# 参照渡し

my $val12 = "ABCD";
test12(\$val12); # \を付けて参照渡しにする
print $val12."\n"; # 値が変わっている

sub test12 {
    my $arg1 = $_[0];
    $$arg1 = "EFG";
}

# //////////// 配列 ////////////

# 値渡し

my @array1 = (0, 1, 2);
test3(@array1);
print Dumper(\@array1)."\n"; # 値は変わっていない

sub test3 {
    my $arg1 = $_[0];
    $arg1[1] = 9999;
}

# 参照渡し

my @array2 = (0, 1, 2);
test13(\@array2); # \を付けて参照渡しにする
print Dumper(\@array2)."\n"; # 値が変わっている

sub test13 {
    my $arg1 = $_[0];
    $$arg1[1] = 9999;
}

File: Ruby/ValueOrRef.rb
# 値渡し

# 数値
def test1(num)
    num *= 2
    printf("test1 %d\n", num)
end
val1 = 10
test1(val1)
printf("%d\n", val1)

# 文字列
def test2(str)
    str += "ZZZ"
    printf("test2 %s\n", str)
end
val2 = "ABC"
test2(val2)
printf("%s\n", val2)

# 文字列 [破壊的メソッド実行時]
def test12(str)
    str.upcase!
    printf("test12 %s\n", str)
end
val12 = "abc"
test12(val12)
printf("%s\n", val12) # 元の値も変更されABCとなる。Rubyではオブジェクトとしてメソッドに引数を渡すため。

# 配列（参照型）
def test3(ary)
    ary[0] = 999
    printf("test3")
    p ary
end
val3 = [1, 2, 3, 4, 5]
test3(val3)
p val3

# ハッシュ（参照型）
def test4(hash)
    hash["Honda"] = 999
    printf("test4")
    p hash
end
val4 = {"Honda" => 1, "Yamaha" => 2, "Kawasaki" => 3}
test4(val4)
p val4

# クラス（参照型）
class User
    @name = ""
    def name
        @name
    end
    def name=(name)
        @name = name
    end
end
def test5(obj)
    obj.name = "Yamaha"
    printf("test5 %s\n", obj.name)
end
val5 = User.new()
val5.name = "Honda"
test5(val5)
printf("%s\n", val5.name)

